### PR TITLE
Anomaly detection sample implementation

### PR DESCRIPTION
**Pull request by AmitRoushan**
**What type of PR is this?**
/kind enhancement

**What this PR does / why we need it**:
The PR is an example implementation of anomaly detection with delfin<->kafka<->prediction pipeline

### CODE REVIEW

File: AnomalyDetection/cmd/data_parser.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+from anomaly_detection import log
+from anomaly_detection.data_parser import manager
+from anomaly_detection.utils import config as cfg
+# need register global_opts
+from anomaly_detection.common import options
+
+CONF = cfg.CONF
+
+data_parser_opts = [
```
Comment (NajmudheenCT) [inline AnomalyDetection/cmd/data_parser.py:24]: Could not find options for kafka reciever..

File: AnomalyDetection/cmd/data_parser.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sys
+
+from anomaly_detection import log
+from anomaly_detection.data_parser import manager
```
Comment (NajmudheenCT) [inline AnomalyDetection/cmd/data_parser.py:17]: Could not see manager module in data praser!!

File: AnomalyDetection/cmd/data_parser.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+               help='Data receiver name')
+]
+
+CONF.register_opts(data_parser_opts, "data_parser")
+
+
+def main():
+    CONF(sys.argv[1:])
+    log.setup(CONF, "anomaly_detection")
+    mgr = manager.Manager(CONF.data_parser.receiver_name)
```
Comment (NajmudheenCT) [inline AnomalyDetection/cmd/data_parser.py:36]: kafka reciever part implementation is missing?

File: AnomalyDetection/cmd/data_parser.py (hunk 1 above)
Comment (AmitRoushan) [reply AnomalyDetection/cmd/data_parser.py]: One commit missed. Squashed commits and updated

File: AnomalyDetection/cmd/data_parser.py (hunk 2 above)
Comment (AmitRoushan) [reply AnomalyDetection/cmd/data_parser.py]: One commit missed. Squashed commits and updated

File: AnomalyDetection/cmd/data_parser.py (hunk 3 above)
Comment (AmitRoushan) [reply AnomalyDetection/cmd/data_parser.py]: One commit missed. Squashed commits and updated

File: DelfinAnomalyDetection/Readme.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+- Delfin publish metric data on kafka (Default topic: delfin-kafka)
+- Anomaly detection module subscribe for the topic and store the metric data in dictionary
+- Anomaly detections runs on stored data in dictionary
+
+### Constraints
+
+- Currently only "read_bandwidth" considered
+- (TODO) Online model training is not implemented 
+- Only 1000 samples are considered for model training
```
Comment (NajmudheenCT) [inline DelfinAnomalyDetection/Readme.md:20]: Can we mention first 1000 samples are considered for modelling and after that the incoming samples are used for prediction

File: DelfinAnomalyDetection/Readme.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+### Steps:
+- Setup Delfin and Kafka with help of [documentation](https://docs.sodafoundation.io/guides/developer-guides/delfin/)
+- install dependencies mentioned in requirements.txt ```pip install -r requirements.txt```
+- export PYTHONPATH=`pwd`
```
Comment (NajmudheenCT) [inline DelfinAnomalyDetection/Readme.md:25]: Hope the steps can be little more detail , like
git clone or download repo
cd /<delfinanalomalidetection> etc??

File: DelfinAnomalyDetection/etc/anomaly_detection.conf [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+log_rotate_interval = 4
+max_logfile_count = 10
+max_logfile_size_mb = 200
+log_rotation_type = "interval"
+logging_default_format_string = "%(asctime)s - %(levelname)s - %(filename)s[:%(lineno)d] - %(message)s"
+#debug = true
+
+[data_parser]
+receiver_name=kafka
+csv_file_name=performance.csv
```
Comment (NajmudheenCT) [inline DelfinAnomalyDetection/etc/anomaly_detection.conf:12]: hope this is not used!

Comment (AmitRoushan) [reply DelfinAnomalyDetection/etc/anomaly_detection.conf]: removed as per suggestion

File: DelfinAnomalyDetection/Readme.md (hunk 5 above)
Comment (AmitRoushan) [reply DelfinAnomalyDetection/Readme.md:28]: added steps for git clone and stepping into directory

File: DelfinAnomalyDetection/Readme.md (hunk 4 above)
Comment (AmitRoushan) [reply DelfinAnomalyDetection/Readme.md]: updated as per comment

**Review by NajmudheenCT — DISMISSED** [review]
LGTM

**Comment by skdwriting** [comment]
PLease provide intro, name the file and provide other needed info

**Comment by AmitRoushan** [comment]
Updated documentation with Intro and Summary.

File: DelfinAnomalyDetection/DelfinKafkaAIBasicPoC.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+| Delfin  | ------>  |  Kafka  |  ----->  | Anomaly Detection |
+ ---------            ---------            -------------------
+````
+
+- Delfin publish metric data on kafka (Default topic: delfin-kafka)
+- Anomaly detection module subscribe for the topic and store the metric data in dictionary
+- Anomaly detections runs on stored data in dictionary
+
+### Steps:
+- Setup Delfin and Kafka with help of [documentation](https://docs.sodafoundation.io/guides/developer-guides/delfin/)
```
Comment (skdwriting) [inline DelfinAnomalyDetection/DelfinKafkaAIBasicPoC.md:23]: This link does not give setup info for delfin and kafka!

Comment (AmitRoushan) [reply DelfinAnomalyDetection/DelfinKafkaAIBasicPoC.md]: Updated link with https://github.com/sodafoundation/delfin/tree/master/installer#test-the-running-delfin-setupprocess

**Review by skdwriting — APPROVED** [review]
LGTM

**Review by NajmudheenCT — APPROVED** [review]

**Pull request merged**